rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala

```diff
@@ -0,0 +1,987 @@
+                .updated(byteToInt(insPrefixRest.head), Leaf(insPrefixRest.tail, insValue))
+              saveNodeAndCreateItem(newNode, commPrefix, compaction = false).some.pure
+            }
+        }
+      }
+    }
+
+    /**
+      * Delete leaf value from this part of tree (start from curItem).
+      * Rehash and save all depend node to cacheW.
+      *
+      * If not found leaf with  delPrefix - return [[None]].
+      * @return Updated current item.
+      */
+    def delete(curItem: Item, delPrefix: ByteVector): F[Option[Item]] = {
+      def deleteFromChildNode(
+          childPtr: ByteVector,
+          childPrefix: ByteVector,
+          delPrefix: ByteVector
+      ): F[Option[Item]] =
+        for {
+          childNode                   <- loadNode(childPtr)
+          (delItemIdx, delItemPrefix) = (byteToInt(delPrefix.head), delPrefix.tail)
+          childItemOpt                <- delete(childNode(delItemIdx), delItemPrefix)
+          newChildNode = childItemOpt.map { childItem =>
+            saveNodeAndCreateItem(childNode.updated(delItemIdx, childItem), childPrefix)
+          }
+        } yield newChildNode
+
+      Sync[F].defer {
+        curItem match {
+          case EmptyItem => none[Item].pure // Not found
+
+          case Leaf(leafPrefix, _) =>
+            if (leafPrefix == delPrefix) (EmptyItem: Item).some.pure // Happy end
+            else none[Item].pure                                     // Not found
+
+          case NodePtr(ptrPrefix, ptr) =>
+            val (commPrefix, delPrefixRest, ptrPrefixRest) = commonPrefix(delPrefix, ptrPrefix)
+            if (ptrPrefixRest.nonEmpty || delPrefixRest.isEmpty) none[Item].pure // Not found
+            else deleteFromChildNode(ptr, commPrefix, delPrefixRest)             // Deeper
+        }
+      }
+    }
+
+    /**
+      * Parallel processing of [[HistoryAction]]s in this part of tree (start from curNode).
+      *
+      * New data load to [[cacheW]].
+      * @return Updated curNode. if no action was taken - return [[None]].
+      */
+    def makeActions(curNode: Node, actions: List[HistoryAction]): F[Option[Node]] = {
+
+      // If we have 1 action in group.
+      // We can't parallel next and we should use sequential traversing with help update() or delete().
+      def processOneAction(action: HistoryAction, item: Item, itemIdx: Int) =
+        for {
+          newItem <- action match {
+                      case InsertAction(key, hash) => update(item, ByteVector(key).tail, hash.bytes)
+                      case DeleteAction(key)       => delete(item, ByteVector(key).tail)
+                    }
+        } yield (itemIdx, newItem)
+
+      def clearingDeleteActions(actions: List[HistoryAction], item: Item) = {
+        val notExistInsertAction = actions.collectFirst { case _: InsertAction => true }.isEmpty
+        if (item == EmptyItem && notExistInsertAction) actions.collect {
```

Comment:
Just curious. When `notExistInsertAction` is `True`, `actions` should be all `DeleteAction`, why do you collect `InsertAction` here?

Reply:
It's my mistake, thanks!